poll: Implement poll notifications in core

jamescowens opened the pull request:
This PR implements the ability to specify an external command to run via -pollnotify=\<cmd\> argument that takes two placeholders, %s1 and %s2, which are replaced with the poll txid and the notification status (added, deleted, or expiration_warning).

The notifications are sent on all non-expired polls. Add/delete notifications are executed by the poll contract handlers. The expiration warning is handled by a scheduler function which checks every 5 minutes and does a single shot notification for polls that are within -pollexpirewarningtime=\<hours\> of expiration.

This should be sufficient to drive an external script for notification via email or other desired approach. The external script can call gridcoinresearchd getpollresults \<poll txid\> or listpolls to get the details of the poll.

This PR includes an example script, poll_notify.sh, to email the poll notification that can be found in the contrib directory.

Closes #2803 and https://github.com/gridcoin-community/Gridcoin-Tasks/issues/255.

jamescowens committed:
Implement poll notifications in core

This commit implements the ability to specify an external command
to run via -pollnotify=<cmd> argument that takes two placeholders,
%s1 and %s2, which are replaced with the poll txid and the
notification status (added, deleted, or expiration_warning).

The notifications are sent on all non-expired polls. Add/delete
notifications are executed by the poll contract handlers. The
expiration warning is handled by a scheduler function which checks
every 5 minutes and does a single shot notification for polls that
are within -pollexpirewarningtime=<hours> of expiration.

This should be sufficient to drive an external script for
notification via email or other desired approach.

div72 commented:
I mean it's kind of fine now since you implemented it but since the poll notify consumers already need to handle things like 1-block forks, you could've just passed the poll txid and left the extra machinery of type and expire to the consumer.

jamescowens commented:
I get where you are coming from; however, not every person using this will have a sophisticated state machine that tracks polls. Some are just enthusiasts running a command line wallet. On the expire warning, it is more  efficient to implement an internal scheduler to check this rather than having someone either 1. creating a sophisticated external state machine to track poll expirations, and/or 2. poll the wallet to check.

This allows the external script to be much simpler for most uses.

The extra little bit of functionality wasn't that hard, anyway! :)

jamescowens committed:
Provide example poll notification bash script.

jamescowens committed:
Add rpc command testpollnotification

This implements a simple rpc command to test poll notification. It
also makes changes to the poll_notify.sh script and a few other
refinements.

jamescowens commented:
Tested this. Both the test rpc and the expiration warning works.

div72 reviewed (approved):
Tested ACK `5e379a8689b0af90a1fb613e454ee744b7a9f75b`.
